site-src/geps/gep-735.md

```diff
@@ -0,0 +1,170 @@
+# GEP-735: TCP and UDP addresses matching
+
+* Issue: [#735](https://github.com/kubernetes-sigs/gateway-api/issues/735)
+* Status: Provisional
+
+## TLDR
+
+Spec for matching source and destination addresses on L4 APIs.
+
+## Goals
+
+- add matching rules for address to `TCPRoute`
+- add matching rules for address to `UDPRoute`
+- intentionally avoid type definitions that would make it hard to expand later
+
+## Non-Goals
+
+- define rules for port matching
+
+## Introduction
+
+While `TCPRoute` and `UDPRoute` currently support custom matching extensions,
+there is desire among the community to include some "fundamental" matching
+options in the spec that cover the most common requirements. In this GEP we
+request address matching for these APIs in order to support a standard
+for some of the commonplace setups of gateway implementations. Matching is
+intended to be covered for both _source_ and _destination_ to enable a finer
+level of tuning options for L4 traffic routing at a level below the `Gateway`.
+
+## API
+
+The API changes include the following new types:
+
+- `AddressMatch` to indicate the IP for address matching
+- `TrafficMatches` to configure matching according to network address
+
+These types enable the address matching required, with some active
+considerations about how to leave these open ended for later expansion.
+
+### AddressMatch Type
+
+A new `AddressMatch` type provides the targeting mechanism for match inclusion
+of a given network address:
+
+```go
+type AddressMatch struct {
+	// Type of the address, either IPAddress or NamedAddress.
+	//
+	// If NamedAddress is used this is a custom and specific value for each
+	// implementation to handle (and add validation for) according to their
+	// own needs.
+	//
+	// For IPAddress the implementor may expect either IPv4 or IPv6.
+	//
+	// +optional
+	// +kubebuilder:validation:Enum=IPAddress;NamedAddress
+	// +kubebuilder:default=IPAddress
+	Type *AddressType `json:"type,omitempty"`
+
+	// Value of the address. The validity of the values will depend
+	// on the type and support by the controller.
+	//
+	// Examples: `1.2.3.4`, `128::1`, `my-named-address`.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	Value string `json:"value"`
+}
+```
```

Comment:
Have you considered making `AddressMatch` a [union type](https://github.com/kubernetes/enhancements/blob/0e4d5df19d396511fe41ed0860b0ab9b96f46a2d/keps/sig-api-machinery/1027-api-unions/README.md), with `Type` being the discriminator?  Having a generic `Value` field looks odd in a Kubernetes API, and it forfeits the opportunity to enforce stricter validation.

Reply:
I had not as I was very intentionally trying to make this similar to `GatewayAddress` so that we didn't end up with two different ways of describing an address, do you have any thoughts on that?

Reply:
Ah, I can see how making `AddressMatch` similar to `GatewayAddress` could reduce confusion if the user wanted to specify the same IP address or named address on both the listener and an address match.  On the other hand, `AddressMatch` is missing `Hostname`, the semantics of `NamedAddress` across the two types gets a little funny, and I'm wary of making things appear alike superficially when they are fundamentally different.  

Elaborating on what I mean by funny semantics, do the names that the user can specify for a `NamedAddress` in the listener refer to the same set of things as the names that the user can specify for a `NamedAddress` in a source match?  What about for destination matches  Is this relationship (or lack thereof) entirely up to the implementation to define?  

(I now kind of wish we had used a union type for `GatewayAddress`, but it's probably too late to change that now without an extremely compelling need.)

Reply:
For better or worse, we've used the `Type` and `Value` pair heavily throughout the API, including path, header, and query param matching. I believe HTTPRoute filters are the only place we've used a proper union type. Although union types are common in upstream Kubernetes, I think the pattern proposed here is also quite common. 

Since we're dealing with CRDs and there is not proper union support yet, we'd be stuck with adding additional webhook validation if we transitioned to a union type. My personal preference would be to continue with the proposed approach to more closely match the pattern we've used elsewhere in Gateway API.

Reply:
Using union types probably would have made sense in those other contexts (and in hindsight, it would have simplified defining support levels).  It makes even more sense for `AddressMatch` because we can specify OpenAPI validation to enforce that any value that the user puts in, say, the "address" field must be either an IPv4 or an IPv6 address (the following is incomplete and not tested, but this example should illustrate what I mean):
```yaml
sourceAddresses:
  items:
    oneOf:
    - properties:
        address:
          format: ipv4
    - properties:
        address:
          format: ipv6
    properties:
      address:
        type: string
    type: object
  type: array
```
Granted, CRD validation may have gaps (if I recall correctly, it admits an object definition that specifies a field that doesn't correspond to the discriminator value).  However, I figure that some validation is better than none.  

I think we'd need webhook validation anyway to validate that the user provided a valid regexp for the path, header, and query param matching, but regexp support is "Custom" and underspecified anyway, so it's fair to leave that up to implementors.  

Using a union type is just a suggestion for enabling slightly improved validation and clarity, so I won't press it if the goal is to make `AddressMatch` match `GatewayAddress`.

Reply:
I think the problem with both using a union type and having more interesting OpenAPI validation is that neither fits well through the CRD-validation-shaped hole we have available for now.

I also think that having some more complex validation in either the webhook or an exported library (as we've discussed briefly a few times) would be beneficial.

Reply:
@Miciah just wanted to double check this?

Reply:
I agree, please disregard the suggestion to use a union type as far as this PR is concerned.